test/cmocka/src/notifier_mocks.c

```diff
@@ -0,0 +1,102 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2019 Intel Corporation. All rights reserved.
+//
+// Author: Artur Kloniecki <arturx.kloniecki@linux.intel.com>
+
+#include <stdint.h>
+#include <stdlib.h>
+#include <sof/lib/alloc.h>
+#include <sof/lib/notifier.h>
+#include <sof/list.h>
+#include <sof/common.h>
+
+struct callback_handle {
+	void *receiver;
+	void *caller;
+	void (*cb)(void *arg, enum notify_id, void *data);
+	struct list_item list;
+};
+
+static struct notify *_notify;
+
+struct notify **arch_notify_get(void)
+{
+	int i;
+
+	if (!_notify) {
+		_notify = malloc(sizeof(*_notify));
+
+		for (i = 0; i < NOTIFIER_ID_MAX; i++)
+			list_init(&_notify->list[i]);
+	}
+
+	return &_notify;
+}
+
+void notifier_event(void *caller, enum notify_id type, uint32_t core_mask,
+		    void *data, uint32_t data_size)
+{
+	struct notify *notify = *arch_notify_get();
+	struct list_item *wlist;
+	struct list_item *tlist;
+	struct callback_handle *handle;
+
+	list_for_item_safe(wlist, tlist, &notify->list[type]) {
+		handle = container_of(wlist, struct callback_handle, list);
+		if (!caller || !handle->caller || handle->caller == caller)
+			handle->cb(handle->receiver, type, data);
+	}
+}
+
+int notifier_register(void *receiver, void *caller, enum notify_id type,
```

Comment:
Not following the API usage here, like why do we need receiver and caller.

Reply:
All callbacks of given type are stored in separate lists.
Receiver is used:
- as arg passed back to the callback when event is triggered
- to match with receivers during unregister call, to only remove specified receiver from the list
Caller is used:
- to determine if triggered event should be propagated through given callback, so that filtering of events is done in notifier, not in receiver
- to match specific callback to remove via unregister

Reply:
I'm aligned with the need for nesting and multiple reiceivers here, but the notifier is intended for notification of events. You are doing a messaging API which should have APIs that look more like Zephyr.

Reply:
@akloniex This would be a lot better (as a notifying - not messaging API) if the producer and consumer had seperate registration APIs. e.g.
```
int notifier_consumer_register(void *consumer, type, callback);
int notifier_producer_register(void *producer, type);
```

Reply:
@lgirdwood In my model, producers don't need to register at all. Maybe the confusion comes from lack of comments, or bad naming used.
Only consumers need to register, and the 'caller' parameter passed to registration is used to indicate to which producer we wish to register.
I could add some elaborate comments to these functions, if that would help to address the confusion.

Reply:
ok, I'm thinking what happens if a reciever registers for a producer that does not exist (say not built in Kconfig option). We would continue atm, but would never be notified of the event.....I can live with this though.

The arg naming though should be more like:
```
notifier_register(type, callback, data);
```
Where callback is function and data is data passed to callback AND used for unregister. I am still not following why we need to pass the void * consumer and how it;s related to propogation.

Reply:
@lgirdwood Think of it this way:
all events of given type, say buffer produce, are multiplexed via single `notify_id`;
there are many buffers in the topology and all contribute to the same event stream;
we might either propagate all these events to all consumers, forcing them to filter the events and react only to those that are relevant to them, or do the filtration in the notifier, based on the pointer to the producer, which was registered by consumer.

I went with the latter option, which saves at least a function call per consumer, but most often more than that, since we usually assign values to variables during declaration in the callback function, which is all wasted time if ultimately received event is not important for given consumer.

Regarding issue of registering to a producer that does not exist yet, that you mentioned:
there is not way of registering to a _specific_ producer when it does not exits, no reasonable deferred method or anything, so I believe we can skip the _specific_ part;
consumer might still register to any event type, with no specific producer in mind (passing `caller == NULL` to register function), effectively listening to all events that land under that `notify_id`, and then deduce if received event is relevant and take appropriate actions.

Reply:
Thanks, I now see where you are going. This sounds like we are going to be broadcasting frquent events from multiple sources ? i.e. the "buffer produce" notification. It would be far better to use a more targeted notification here i.e. buffer X "produce" -> consumer Y, this needs a little bit more logic to setup but scales well.

The notification framework is more intended for low frequency broadcast events (e..g changing clock freq) and not suited for high frequency broacasting (i.e. the buffer produce event). This needs to be more targetted for high frequency events. 

Can you describe the event you are targetting here. Whats the use case.

Reply:
I had 2 target applications in mind, both regarding buffer callbacks:
- probes
- echo reference
Both of which require support for multiple callbacks attached to one instance of buffer.

DMA callbacks got onto the train of changes mostly to unify usage of all callback types across the codebase.

Could you describe how would you like to implement the X->Y correspondence?
I believe that, for the most part, my solution gives that possibility, minimizing footprint and also giving the possibility to broadcast events. I'm kind of biased towards my solution since I've already written it, and can't think of a better way to implement this, but I'm open for fresh ideas on the subject.

Reply:
So for both probes and echo reference we need to "messge by reference". i.e. "probe X is attatched to BUFFER1.0" by FDK, or "echo ref subscribes to BUFFER2.0 produce" set by topology config for ech ref.
The reference part can be a string based on module/comp name and module/comp ID (similar to CAVS FW).

Reply:
@akloniex apologies for the delay. As discussed, the component/module messaging is different wrt to targetting and freqency with general system notifications. I think we need to define.
1) A set of component messages. i.e. "buffer produce", "buffer consume", "component stop", etc and add them as topology tuple tokens attributes for modules/components.
 2) Create an insertion macro so they can be added as attributes in topology components. This macro should also embed the source/target component ID.
3) Add light hooks into pipeline, buffer and components to send the messages. The messaging  mappings will be bound at pipeline creation time, i..e no need to search and map at runtime.

Reply:
@lgirdwood The issue I have with your approach is that probes, which are one of most important users of this, are not specified in topology at all, and uapi for them was already specified and passed down form cAVS.
Also, I don't see that alleviating the hassle of binding multiple consumers to single message. How do you see that part solved?

Reply:
I would class probes as a system debug user with direct access to buffer (including methods in the buffer structures). IPC will use the buffer ID and bind it to a SW probe.
Are there any other use cases for multiple consumers to a single user (aside from our current system uses today).

Reply:
@lgirdwood I don't understand why you want to have 1-1 relationship for events. For me it's counter-intuitive. I'd bet it's counter-intuitive for most of people as I usually see event notifiers to be always implemented as 1-to-many in various SDKs (when we consider one central notifier). What you want to force here is some kind of hard bind, but it shouldn't be named notifier then, but something more like observer+observable.